apps/cowswap-frontend/src/modules/bridge/pure/ProtocolIcons/index.tsx

```diff
@@ -10,63 +12,215 @@ export interface ProtocolIconsProps {
-      </ProtocolIcon>
-    )
+// Utility function for mask configuration
+const createMask = (
+  baseCx: number,
+  offsetMultiplier: number,
+  distance: number,
+  centerY: number,
+  innerR: number,
+  outerR: number,
+): MaskConfig => ({
+  cx: baseCx + offsetMultiplier * distance,
+  cy: centerY,
+  innerR: innerR,
+  outerR: outerR,
+})
+
+// Custom hook for mask calculations. Memoized only for costly computation.
+function useMaskCalculations(
+  currentDisplaySize: number,
+  hoveredIcon: 'first' | 'second' | null,
+): MaskCalculationResult {
+  return useMemo(() => {
+    const cutThickness = getBorderWidth(currentDisplaySize)
+    const radiusOfTopIcon = currentDisplaySize / 2
+    const maskOuterRadius = radiusOfTopIcon + cutThickness
+    const maskCenterY = currentDisplaySize / 2
+    const iconCenterBaseX = currentDisplaySize / 2
+    const distanceBetweenCenters = currentDisplaySize * (1 - STACKED_ICON_OVERLAP_RATIO)
+
+    const isFirstHovered = hoveredIcon === 'first'
+    const isSecondHovered = hoveredIcon === 'second'
+    const isFirstOnTop = isFirstHovered
+    const isSecondOnTop = isSecondHovered || (!isFirstHovered && !isSecondHovered)
+
+    const firstMaskConfig = !isFirstOnTop
+      ? createMask(iconCenterBaseX, 1, distanceBetweenCenters, maskCenterY, radiusOfTopIcon, maskOuterRadius)
+      : undefined
+    const secondMaskConfig = !isSecondOnTop
+      ? createMask(iconCenterBaseX, -1, distanceBetweenCenters, maskCenterY, radiusOfTopIcon, maskOuterRadius)
+      : undefined
+
+    return {
+      firstMaskConfig,
+      secondMaskConfig,
+      isFirstOnTop,
+      isSecondOnTop,
+    }
+  }, [currentDisplaySize, hoveredIcon])
+}
+
+// Single Protocol Icon Component
+interface SingleProtocolIconProps {
+  showOnlyFirst?: boolean
+  secondProtocol: BridgeProviderInfo
+  currentDisplaySize: number
+  currentLogoHeight: number
+}
+
+const SingleProtocolIcon = memo(function SingleProtocolIcon({
+  showOnlyFirst,
+  secondProtocol,
+  currentDisplaySize,
+  currentLogoHeight,
+}: SingleProtocolIconProps): React.JSX.Element {
+  const protocolName = showOnlyFirst ? 'CoW Swap' : secondProtocol.name
+  const protocolBgColor = showOnlyFirst ? UI.COLOR_BLUE_300_PRIMARY : undefined
+  const iconChild = showOnlyFirst ? (
+    <ProductLogo variant={ProductVariant.CowSwap} height={currentLogoHeight} logoIconOnly />
+  ) : (
+    <img src={secondProtocol.logoUrl} width={currentLogoHeight} height={currentLogoHeight} alt={secondProtocol.name} />
+  )
+
+  return (
+    <ProtocolIcon title={protocolName} size={currentDisplaySize} isStacked={false} bgColor={protocolBgColor}>
+      {iconChild}
+    </ProtocolIcon>
+  )
+})
+
+// Stacked Protocol Icons Component
+interface StackedProtocolIconsProps {
+  secondProtocol: BridgeProviderInfo
+  currentDisplaySize: number
+  currentLogoHeight: number
+}
+
+const StackedProtocolIcons = memo(function StackedProtocolIcons({
```

Comment:
This component deserves it's own file.
Please try to not make component files with more than ~150 lines of code.

Reply:
Addressed.